File: src/Libraries/CoreLib/Collections/Stack.cs
// Stack.cs
// Script#/Libraries/CoreLib
// Copyright (c) Nikhil Kothari.
// Copyright (c) Microsoft Corporation.
// This source code is subject to terms and conditions of the Microsoft 
// Public License. A copy of the license can be found in License.txt.
//

using System;
using System.Runtime.CompilerServices;

namespace System.Collections {

    /// <summary>
    /// The Stack data type which is mapped to the Array type in Javascript.
    /// </summary>
    [IgnoreNamespace]
    [Imported]
    [ScriptName("Array")]
    public sealed class Stack {

        [IntrinsicProperty]
        [ScriptName("length")]
        public int Count {
            get {
                return 0;
            }
        }

        public void Clear() {
        }

        public bool Contains(object item) {
            return false;
        }

        public object Peek() {
            return null;
        }

        public object Pop() {
            return null;
        }

        public void Push(object item) {
        }
    }
}

File: src/DSharp.Script/src/System/Collections/Stack.js
function Stack() {
    this.count = 0;
    this._items = [];
}
var Stack$ = {
    clear: function () {
        this._items.length = 0;
        this.count = 0;
    },
    contains: function (item) {
        for (var i = this.count - 1; i >= 0; i--) {
            if (this._items[i] === item) {
                return true;
            }
        }
        return false;
    },
    getEnumerator: function () {
        return new Enumerator(this._items.reverse());
    },
    peek: function () {
        return this._items[this.count - 1];
    },
    push: function (item) {
        this._items.push(item);
        this.count++;
    },
    pop: function () {
        if (this.count) {
            this.count--;
            return this._items.pop();
        }
        return undefined;
    }
};